service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/SendAcknowledgementExecutor.java

```diff
@@ -39,25 +48,66 @@
     @SneakyThrows
     public void execute(SendAcknowledgementTaskDefinition sendAcknowledgementTaskDefinition) {
         LOGGER.info("Sending application acknowledgement from the adaptor to the requesting system");
+        var ackMessageId = randomIdGeneratorService.createNewId();
 
         var sendAckTemplateParams = SendAckTemplateParameters.builder()
+            .requestId(ackMessageId)
             .creationTime(DateFormatUtil.toHl7Format(timestampService.now()))
-            .uuid(randomIdGeneratorService.createNewId())
+            .messageId(sendAcknowledgementTaskDefinition.getEhrRequestMessageId())
             .fromAsid(sendAcknowledgementTaskDefinition.getFromAsid())
             .toAsid(sendAcknowledgementTaskDefinition.getToAsid())
             .typeCode(sendAcknowledgementTaskDefinition.getTypeCode())
-            .messageId(sendAcknowledgementTaskDefinition.getEhrRequestMessageId())
+            .reasonCode(sendAcknowledgementTaskDefinition.getReasonCode())
+            .reasonMessage(sendAcknowledgementTaskDefinition.getDetail())
             .build();
-        var acknowledgementRequestBody = TemplateUtils.fillTemplate(ACKNOWLEDGEMENT_TEMPLATE, sendAckTemplateParams);
-        var outboundMessage = OutboundMessage.builder().payload(acknowledgementRequestBody).build();
-        var stringRequestBody = objectMapper.writeValueAsString(outboundMessage);
-        var positiveAckMessageId = randomIdGeneratorService.createNewId();
 
-        var request = mhsRequestBuilder.buildSendAcknowledgement(stringRequestBody, sendAcknowledgementTaskDefinition.getFromOdsCode(),
-            sendAcknowledgementTaskDefinition.getConversationId(), positiveAckMessageId);
+        String requestBody = buildRequestBody(sendAckTemplateParams);
+
+        sendToMHS(sendAcknowledgementTaskDefinition, ackMessageId, requestBody);
+
+        updateEhrExtractStatus(sendAcknowledgementTaskDefinition, ackMessageId);
+    }
+
+    private void sendToMHS(SendAcknowledgementTaskDefinition taskDefinition, String ackMessageId, String requestBody) {
+        LOGGER.info("Sending ACK message to MHS. ACK message Id: {} Conversation id: {} EhrRequest id {} ", ackMessageId,
+            taskDefinition.getConversationId(), taskDefinition.getEhrRequestMessageId());
+
+        var request = mhsRequestBuilder.buildSendAcknowledgement(
+            requestBody, taskDefinition.getFromOdsCode(), taskDefinition.getConversationId(), ackMessageId);
 
         mhsClient.sendMessageToMHS(request);
+    }
+
+    private String buildRequestBody(SendAckTemplateParameters templateParams) throws JsonProcessingException {
+        var template = StringUtils.isNoneBlank(templateParams.getReasonCode()) ? NEGATIVE_ACK_TEMPLATE : ACKNOWLEDGEMENT_TEMPLATE;
```

Comment:
Does `reasonCode` always correspond to an error?

Reply:
yes, for positive ack it will be always empty as the story says: https://gpitbjss.atlassian.net/browse/NIAD-840 but I see it would be more meaningful to check it with the typeCode.